### PR TITLE
Finalized upload method and redefined modes (using an enum)

### PR DESCRIPTION
**Pull request by jlumbroso**
- Finalized `upload_submission` following to conversation with @jamesaevans (granularity of aux method with add/change; remove comments + unclaim without deleting in overwrite mode; only make mods if changes are made to the files).
- Redefined all upload modes according to a new set of criteria.
- Used the `Enum` class from Python 3.4+ for better coding practice with the upload modes. They will be compared by name rather than value (useful since they are defined as dicts) and can have docstrings attached to individual modes.

### CODE REVIEW

File: codePost_lib.py [hunk 1]
```diff
@@ -13,42 +13,159 @@
...
+        "updateExistingFiles": True,
+
+        "removeComments": True,
+        "doUnclaim": True,
+        "deleteAffectedSubmissions": True
+    }, """
+    With the 'Overwrite' mode: If a submission already exists for this (student, assignment) pair
+    (including partners), overwrite it with the contents of the request. Keep the existing submission
+    linked to any partners not included in the request. Delete any existing comments, unclaim
+    the submission, and set grader field to `None`.
```
Comment (jamesaevans) [inline codePost_lib.py]: Suggest we include a line here that mentions the fact that a submission will only be updated (unclaimed, comments cleared), if either (a) a new file is added, (b) a file is updated (or both).

File: codePost_lib.py [hunk 2]
```diff
@@ -108,14 +225,14 @@ def upload_submission(api_key, assignment, students, files, mode=DEFAULT_MODE):
...
-        if set(students) != set(existing_submissions[0]["students"]) or len(existing_submissions) > 1:
+        if len(existing_submissions) > 1 or set(students) != set(existing_submissions[0]["students"]):
             raise UploadError(
                 """
                 There are {} existing submission(s) with a different subset of
                 students than those requested. Since 'resolveStudents' is false,
                 interrupting upload.
                 - Requested students: {}
-                - Existing students (on first existing submission): {}
+                - Existing students(on first existing submission): {}
```
Comment (jamesaevans) [inline codePost_lib.py]: Should be `Existing students (on first existing submission)`

Comment (jamesaevans) [reply codePost_lib.py]: Note the deleted space between 's' and '()'

File: codePost_lib.py [hunk 3]
```diff
@@ -474,6 +601,118 @@ def get_assignment_submissions(api_key, assignment_id, username=None):
...
+               {}
+            """.format(submission_id, grader, exc)
+        )
+
+    return False
+
+
+def unclaim_submission(api_key, submission_id):
+    """
+    Unclaims a submission, given the submission ID. This unset the associated
```
Comment (jamesaevans) [inline codePost_lib.py]: Should be `unsets` instead of `unset`

**Comment by jamesaevans** [comment]
What do you think of separating the submission-specific functions from the more general helper functions? In the latter category, we could separate out the following functions into a file:
- `get_available_courses`
- `get_assignment_info_by_id`
- `get_file`
- `get_assignment_info_by_name`
- `get_assignment_submissions`
- `set_submission_grader`
- `unclaim_submission`
- `remove_comments`
- `delete_submission`
- `delete_file`
- `post_file`

To me, separating these files could allow people to better understand how to use the API. Step 1 is understand the basic functions (separated into one file). Step 2 is to compose these simple functions into more complicated, advanced actions (like the submission infrastructure, separated into its own file).

**Comment by jlumbroso** [comment]
That's a good idea to immediately refactor the current file into two files: A `submission` submodule and a `helper` submodule.

I think I would also group the following methods as part of the `submission` submodule:
- `set_submission_grader`;
- `unclaim_submission`;
- `delete_submission`.

But eventually, here is how I see this project evolving:
1. Package this example as a Python package, with a binary CLI tool. (Install with pip.)
2. Embed OpenAPI client to our autogenerated REST API as main interface, and helper methods as "shortcuts."
3. Add testing infrastructure for regression tests.

**Pull request merged**